mvc/models/model-binding.rst

```diff
@@ -1,8 +1,91 @@
+
+`http://contoso.com/movies/edit/2`
+
+Since the route template looks like this, ``{controller=Home}/{action=Index}/{id?}``, ``movies/edit/2`` routes to the ``Movies`` controller, and its ``Edit`` action method. It also accepts an optional parameter called ``id``. The code for the action method should look something like this: 
+
+.. code-block:: c#
+   :dedent: 8
+   :linenos:
+   
+           public IActionResult Edit(int? id)
+   
+.. note:: The strings in the URL route are not case sensitive.
+
+MVC will try to bind request data to the action parameters by name. MVC will look for values for each parameter using the parameter name and the names of its public settable properties. In the above example, the only action parameter is named ``id``, which MVC binds to the value with the same name in the route values. In addition to route values MVC will bind data from various parts of the request and it does so in a set order. Below is a list of the data sources in the order that model binding looks through them:
+ 
+#. ``Form values``: These are form values that go in the HTTP request using the POST method. (including jQuery POST requests).
+#. ``Route values``: The set of route values provided by `routing <https://docs.asp.net/projects/mvc/en/latest/controllers/routing.html>`_. 
+#. ``Query strings``: The query string part of the URI.
+
+.. note:: Form values, route data, and query strings are all stored as name-value pairs.
+
+Since model binding asked for a key named ``id`` and there is nothing named ``id`` in the form values, it moved on to the route values looking for that key. In our example, it's a match. Binding happens, and the value is converted to the integer 2. The same request using Edit(string id) would convert to the string "2". 
+
+So far the example uses simple types. In MVC simple types are any .NET primitive type or type with a string type converter. If the action method's parameter were a class such as the ``Movie`` type, which contains both simple and class types as properties, MVC's model binding will still handle it nicely. It uses reflection and recursion to traverse the properties of class types looking for matches. Model binding looks for the pattern parameter_name.property_name to bind values to properties. If it doesn't find matching values of this form, it will attempt to bind using just the property name. For those types such as ``Collection`` types, model binding looks for matches to `parameter_name[index]` or just `[index]`. Model binding treats  ``Dictionary`` types similarly, asking for `parameter_name[key]` or just `[key]`, as long as they keys are simple types. Keys that are supported match the field names HTML and tag helpers generated for the same model type. This enables round-tripping values so that the form fields remain filled with the user's input for their convenience, for example, when bound data from a create or edit did not pass validation.
```

Comment:
- Believe we normally capitalize "Reflection" in docs.
- Sentence beginning "Keys that are supported ..." feels like the start of a new paragraph. Part of the issue is proximity to the use of `[key]` in the clause about dictionaries. So it may also help to start the new paragraph with "Names that are supported ...".
- Suggest "complex types" for both "class types" above.